tests/unit/accounts/test_utils.py

```diff
@@ -0,0 +1,221 @@
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import base64
+import collections
+
+import pretend
+import pytest
+
+from warehouse.accounts import utils
+
+basic_auth_pypi_1234 = base64.b64encode(b"__token__:pypi-1234").decode("utf-8")
+
+
+def test_token_leak_matcher_extract():
+    with pytest.raises(NotImplementedError):
+        utils.TokenLeakMatcher().extract("a")
+
+
+def test_plain_text_token_leak_matcher_extract():
+    assert utils.PlainTextTokenLeakMatcher().extract("a") == "a"
+
+
+def test_base64_basic_auth_token_leak_extract():
+    assert (
+        utils.Base64BasicAuthTokenLeakMatcher().extract(basic_auth_pypi_1234)
+        == "pypi-1234"
+    )
+
+
+@pytest.mark.parametrize(
+    "input", [base64.b64encode(b"pypi-1234").decode("utf-8"), "foo bar"]
+)
+def test_base64_basic_auth_token_leak_extract_error(input):
+    with pytest.raises(utils.ExtractionFailed):
+        utils.Base64BasicAuthTokenLeakMatcher().extract(input)
+
+
+def test_invalid_token_leak_request():
+    exc = utils.InvalidTokenLeakRequest("a", "b")
+
+    assert str(exc) == "a"
+    assert exc.reason == "b"
+
+
+@pytest.mark.parametrize(
+    "record, error, reason",
+    [
+        (None, "Record is not a dict but: None", "format"),
+        ({}, "Record is missing attribute(s): token, type, url", "format"),
+        (
+            {"type": "not_found", "token": "a", "url": "b"},
+            "Matcher with code not_found not found. "
+            "Available codes are: token, base64-basic-auth",
+            "invalid_matcher",
+        ),
+        (
+            {"type": "base64-basic-auth", "token": "foo bar", "url": "a"},
+            "Cannot extract token from recieved match",
+            "extraction",
+        ),
+    ],
+)
+def test_token_leak_disclosure_request_from_api_record_error(record, error, reason):
+    with pytest.raises(utils.InvalidTokenLeakRequest) as exc:
+        utils.TokenLeakDisclosureRequest.from_api_record(record)
+
+    assert str(exc.value) == error
+    assert exc.value.reason == reason
+
+
+@pytest.mark.parametrize(
+    "type, token",
+    [("token", "pypi-1234"), ("base64-basic-auth", basic_auth_pypi_1234)],
+)
+def test_token_leak_disclosure_request_from_api_record(type, token):
+    request = utils.TokenLeakDisclosureRequest.from_api_record(
+        {"type": type, "token": token, "url": "http://example.com"}
+    )
+
+    assert request.token == "pypi-1234"
+    assert request.public_url == "http://example.com"
+
+
+def test_token_leak_analyzer_analyze_disclosure(monkeypatch):
+
+    metrics = collections.defaultdict(int)
```

Comment:
Did you mean counter?
```suggestion
    metrics = collections.Counter()
```

Reply:
Ah, yes, both work. In this case, I'll do ``metrics.update(key)``.

I'll do it, no worry, but is there a specific reason I should learn to use one over the other ?

Reply:
(done)

Reply:
It's just something that fits better semantically, hence the code would read better.